File: python/src/linodemcp/profiles/builder/match.py
"""Wildcard pattern expansion for the draft builder.

Mirrors ``go/internal/profiles/builder/match.go``. Used by the Phase
8.4 mutator methods on :class:`linodemcp.profiles.builder.Registry`
to turn the model's literal-or-wildcard tool arguments into the
explicit list stored on the Draft.
"""

from __future__ import annotations

import fnmatch
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Sequence

    from linodemcp.profiles.builtin import ToolDescriptor


_WILDCARD_CHAR = "*"


def match_patterns(
    patterns: Sequence[str],
    catalog: Sequence[ToolDescriptor],
) -> list[str]:
    """Expand patterns against the catalog, deduplicate, sort.

    - Literal entries (no ``*``) must equal a catalog tool name to
      contribute. Unknown literals contribute nothing rather than
      raising; the caller reports a hit count so the model can
      detect typos by absence.
    - Wildcard entries use ``fnmatch.fnmatch`` (shell-glob). The
      same name produced by multiple patterns appears once in the
      output.

    Returns an empty list when ``patterns`` is empty; never ``None``.
    """
    seen: set[str] = set()
    out: list[str] = []

    for pattern in patterns:
        if not pattern:
            continue

        for name in _match_one(pattern, catalog):
            if name in seen:
                continue

            seen.add(name)
            out.append(name)

    out.sort()
    return out


def _match_one(pattern: str, catalog: Sequence[ToolDescriptor]) -> list[str]:
    """Expand a single pattern against the catalog."""
    if _WILDCARD_CHAR not in pattern:
        for entry in catalog:
            if entry.name == pattern:
                return [pattern]
        return []

    return [entry.name for entry in catalog if fnmatch.fnmatch(entry.name, pattern)]

File: go/internal/profiles/builder/match.go
package builder

import (
	"path/filepath"
	"slices"
	"strings"

	"github.com/chadit/LinodeMCP/go/internal/profiles"
)

// wildcardChar is the only glob metacharacter the builder honors in
// pattern arguments. Matches the spec used by the active-profile
// resolver so user-facing wildcard semantics stay consistent across
// the builder and config-driven profiles.
const wildcardChar = "*"

// MatchPatterns expands the given list of literal-or-wildcard
// patterns against a tool catalog and returns the deduplicated,
// sorted list of matching tool names. Used by the Phase 8.4
// `_add_tools` and `_remove_tools` handlers.
//
//   - Literal entries (no "*") must equal a registered tool name to
//     contribute. Unknown literals contribute nothing rather than
//     erroring out; the handler reports a hit count so the caller
//     can detect typos by absence.
//   - Wildcard entries use filepath.Match (shell-glob). A malformed
//     pattern matches nothing.
//   - The same name produced by multiple patterns appears once.
//
// Returns an empty slice when patterns is nil or empty; never nil.
func MatchPatterns(patterns []string, catalog []profiles.ToolDescriptor) []string {
	seen := make(map[string]struct{}, len(catalog))
	out := make([]string, 0, len(patterns))

	for _, pattern := range patterns {
		if pattern == "" {
			continue
		}

		matches := matchOne(pattern, catalog)
		for _, name := range matches {
			if _, dup := seen[name]; dup {
				continue
			}

			seen[name] = struct{}{}

			out = append(out, name)
		}
	}

	slices.Sort(out)

	return out
}

// matchOne expands a single pattern against the catalog. Split from
// MatchPatterns to keep the loop body readable; behavior is equivalent
// to the active-profile resolver's matchPattern helper.
func matchOne(pattern string, catalog []profiles.ToolDescriptor) []string {
	if !strings.Contains(pattern, wildcardChar) {
		for i := range catalog {
			if catalog[i].Name == pattern {
				return []string{pattern}
			}
		}

		return nil
	}

	matches := make([]string, 0)

	for idx := range catalog {
		ok, err := filepath.Match(pattern, catalog[idx].Name)
		if err != nil || !ok {
			continue
		}

		matches = append(matches, catalog[idx].Name)
	}

	return matches
}
